daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -852,6 +853,90 @@ private[lf] object SBuiltin {
     }
   }
 
+  final object SBScaleBigNumeric extends SBuiltinPure(1) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SInt64 = {
+      val x = args.get(0).asInstanceOf[SBigNumeric].value
+      SInt64(x.scale().toLong)
+    }
+  }
+
+  final object SBPrecisionBigNumeric extends SBuiltinPure(1) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SInt64 = {
+      val x = args.get(0).asInstanceOf[SBigNumeric].value
+      SInt64(x.precision().toLong)
+    }
+  }
+
+  final object SBAddBigNumeric extends SBuiltinPure(2) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SBigNumeric = {
+      val x = args.get(0).asInstanceOf[SBigNumeric].value
+      val y = args.get(1).asInstanceOf[SBigNumeric].value
+      rightOrArithmeticError("overflow/underflow", SBigNumeric.fromBigDecimal(x add y))
+    }
+  }
+
+  final object SBSubBigNumeric extends SBuiltinPure(2) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SBigNumeric = {
+      val x = args.get(0).asInstanceOf[SBigNumeric].value
+      val y = args.get(1).asInstanceOf[SBigNumeric].value
+      rightOrArithmeticError("overflow/underflow", SBigNumeric.fromBigDecimal(x subtract y))
+    }
+  }
+
+  final object SBMulBigNumeric extends SBuiltinPure(2) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SBigNumeric = {
+      val x = args.get(0).asInstanceOf[SBigNumeric].value
+      val y = args.get(1).asInstanceOf[SBigNumeric].value
+      rightOrArithmeticError("overflow/underflow", SBigNumeric.fromBigDecimal(x multiply y))
+    }
+  }
+
+  final object SBDivBigNumeric extends SBuiltinPure(4) {
+    override private[speedy] def executePure(args: util.ArrayList[SValue]): SBigNumeric = {
+      val scale = rightOrCrash(SBigNumeric.checkScale(args.get(0).asInstanceOf[SInt64].value))
+      val roundingMode =
+        java.math.RoundingMode.valueOf(args.get(1).asInstanceOf[SInt64].value.toInt)
```

Comment:
What happens for invalid scales? Does this crash or default to something weird?

Reply:
If scale is invalid it crash.